### PR TITLE
API-1602: dns lookup for event subscription url

### PR DESCRIPTION
**Pull request by tseho**
**Rules**

Blacklist of the domains:
- localhost
- elasticsearch
- memcached 

Check URL for use of IP addresses instead of FQDN's and blacklist all RFC 1918 IPs

Set a new env variable to allow a coma separated list of specific network. For example: `ALLOWED_NETWORK_WHITELIST=10.0.2.0/24,172.19.8`

**Description (for Contributor and Core Developer)**

In the Event Subscription endpoint field, we want to check 2 things:
First, only urls are allowed (http & https), not other protocols.
Screenshot_2021-06-09_17-10-30
Second, urls resolving to internal ips are forbidden.
Screenshot_2021-06-09_17-09-49

**Documentation's PR:** https://github.com/akeneo/pim-docs/pull/1643

**Definition Of Done (for Core Developer only)**

- [ ] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [x] Tech Doc

### CODE REVIEW

**Review by tomglvng — APPROVED** [review]

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/Symfony/Resources/config/validators.yml [hunk 1]
```diff
@@ -1,3 +1,6 @@
+parameters:
+    env(ALLOWED_NETWORK_WHITELIST): ''
```
Comment (pchasle) [inline src/Akeneo/Connectivity/Connection/back/Infrastructure/Symfony/Resources/config/validators.yml:2]: Should you not add `ALLOWED_NETWORK_WHITELIST` to the `.env` file too? It's easier to see what env vars exists when they are all defined in one place.

**Review by pchasle — APPROVED** [review]
Should we prepare documentation for the `6.0` on https://docs.akeneo.com/master/install_pim/manual/events_api.html about `ALLOWED_NETWORK_WHITELIST`?

**Comment by tseho** [comment]
I've already opened a PR for the doc https://github.com/akeneo/pim-docs/pull/1643

**Comment by pchasle** [comment]
> I've already opened a PR for the doc [akeneo/pim-docs#1643](https://github.com/akeneo/pim-docs/pull/1643)

Great :pray:

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/Symfony/Resources/config/validators.yml (hunk 1 above)
Comment (tseho) [reply src/Akeneo/Connectivity/Connection/back/Infrastructure/Symfony/Resources/config/validators.yml:2]: I hesitate, I don't know if we should pollute the `.env` with an option that will not be used in 99% of the cases.

File: src/Akeneo/Connectivity/Connection/back/tests/EndToEnd/Webhook/UpdateEventSubscriptionEndToEnd.php [hunk 2]
```diff
@@ -44,7 +44,7 @@ public function test_it_updates_an_event_subscription(): void
         $data = [
             'code' => $connection->code(),
             'enabled' => true,
-            'url' => 'http://localhost',
+            'url' => 'http://example.com',
```
Comment (pchasle) [inline src/Akeneo/Connectivity/Connection/back/tests/EndToEnd/Webhook/UpdateEventSubscriptionEndToEnd.php:47]:
```suggestion
            'url' => 'http://example.test',
```
It's minor, but if you can, use the `.test` domain extension that was created/reserved for this purpose (and you will never risk to call a real domain)

Comment (pchasle) [reply src/Akeneo/Connectivity/Connection/back/tests/EndToEnd/Webhook/UpdateEventSubscriptionEndToEnd.php:47]: In this case, if `FakeDnsLookup.php` were to be removed the test will probably pass, and call `example.com` for real.

Comment (tseho) [reply src/Akeneo/Connectivity/Connection/back/tests/EndToEnd/Webhook/UpdateEventSubscriptionEndToEnd.php:47]: Fine to me, anyway, I was not resolving the DNS in the tests, it's mocked, even for e2e.

File: src/Akeneo/Connectivity/Connection/back/Application/Webhook/Validation/ExternalUrlValidator.php [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+/**
+ * @copyright 2021 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class ExternalUrlValidator extends ConstraintValidator
+{
+    private const DOMAIN_BLACKLIST = [
+        'localhost',
+        'elasticsearch',
+        'memcached',
```
Comment (StevenVAIDIE) [inline src/Akeneo/Connectivity/Connection/back/Application/Webhook/Validation/ExternalUrlValidator.php:22]: Should we also add object-storage & mysql ?

```suggestion
        'memcached',
        'object-storage',
        'mysql',
```

**Review by StevenVAIDIE — APPROVED** [review]

**Review by StevenVAIDIE — COMMENTED** [review]
~~Do you need to check that protocol url is http / https ? Not file for example~~ => sorry i didn't look the ticket

**Comment by tseho** [comment]
> sorry i didn't look the ticket

It's ok, also, we were already validating that it was a valid url (see my first screenshot), `ftp:/` and `file:/` are already forbidden.

File: src/Akeneo/Connectivity/Connection/back/Application/Webhook/Validation/ExternalUrlValidator.php (hunk 3 above)
Comment (tseho) [reply src/Akeneo/Connectivity/Connection/back/Application/Webhook/Validation/ExternalUrlValidator.php:22]: I've added the question to the ticket, waiting for clarification.

**Pull request merged**